### ISSUE TITLE
Warn users about common Patrol misconfiguration problems

### ISSUE THREAD

**Issue by bartekpacia** [leancodepl/patrol#1193]
### The problem

Developers often have problems with setting up Patrol in their apps.

### Possible solutions

Because of the complexity and dynamic nature of mobile development tooling (Gradle build scripts, a myriad of ways to configure Xcode projects), it's impossible to introduce a `patrol bootstrap`-like command that would perform the set up. 

What we can try doing instead is developing **heuristics-based checks** that would warn developers about common Patrol misconfiguration problems. When a developer were to run `patrol test` (or `patrol build`, or any other command that requires native setup), the checks would be performed.

### Example

``
### List of Android heuristics

1. File `MainActivityTest.java` not found

Inputs: app package name (`com.example.myapp`) - we could get it from the `patrol` section in `pubspec.yaml`
Action: check if file `MainActivityTest.java` exists in `android/app/src/androidTest/java/com/example/myapp`

### List of iOS heuristics

1. Target `RunnerUITests` doesn't exist

Inputs: none
Actions: check if  file `ios/RunnerUITests/RunnerUITests.m` exists

### Alternatives 

Another way to reduce configuration-related churn would be to [make `patrol develop` not require configuration at all](https://github.com/leancodepl/patrol/issues/1030#issuecomment-1455886284). Native configuration would still be required to get test reports and to run on devicefarms, though.

**Comment by bartekpacia** [comment]
Please comment on this issue to add ideas for more heuristics.

**Comment by Albert221** [comment]
> When a developer were to run patrol test (or patrol build, or any other command that requires native setup), the checks would be performed.

Yes, absolutely. I think that would reduce the number of people that don't understand why native integration doesn't work significantly while still offering a good UX in the CLI

**Comment by bartekpacia** [comment]
> Maybe for more complex projects it would indeed be very tough, and then we could show a message that a setup requires manual work. But for simpler projects (that don't have any changes in their native files or if some basic criteria are met) such bootstraping could be helpful, for fresh projects (like PoC, checking how patrol works) or for basic projects, created by less experienced devs (that would greater benefit from such bootstrap)

That'd be hard to develop, and hard to test across different Flutter versions.

I prefer no automatic `patrol bootstrap` at all than a `bootstrap` that is flaky and doesn't work except for fresh, `flutter create`d app. Then the advice we'd give to our users, who will complain that "`patrol bootstrap` doesn't work", would probably be "don't use `patrol bootstrap`', and it doesn't look good.

Also, I'm pretty certain we don't have the resources to maintain the complexity that such `patrol bootstrap` feature would add.

**Comment by neiljaywarner** [comment]
@bartekpacia  i am not sure if this warrants another issue, but i'll be honest, at this point I'm confused whether or not we can use patrol WITHOUT native automatin in *all* widget and integration tests, ie 
1) with flutter test instead of patro ltest
2) with coverage
3) with.. well. i hate to say it but "less issues". we are having some problems still where tests pass in patrol test but fail in flutter logs

**Comment by bartekpacia** [comment]
Hi @neiljaywarner,

Patrol has 2 features: _custom finders_ and _native automation_. You can use one of them, or both.

If you want to only use _custom finders_ ("the dollar" – `$`), you can definitely do this. What problem are you experiencing?


>  we are having some problems still where tests pass in patrol test but fail in flutter logs

This sounds like a bug - and one that we don't experience ourselves.

**Comment by bartekpacia** [comment]
Minimal version: create a page in docs about these common problems

**Comment by fylyppo** [comment]
Maybe it'll be good to warn users about incompatible:
* Gradle + Java versions https://docs.gradle.org/current/userguide/compatibility.html
* Java class version e.g when you try to load something compiled for Java 6 in a Java 5 runtime
